### PR TITLE
CloudMonitoring: Move data manipulation to backend

### LINKED ISSUE #41377: CloudMonitoring: move data processing to backend
part of #39907

Move the code for processing the resource data into backend and use CheckHealth for testing the datasource.

### PR DESCRIPTION
**Pull request by iwysiu**
**What this PR does / why we need it**:
Moves the data manipulation for the resource calls to the backend.

**Which issue(s) this PR fixes**:

Fixes #41377 

**Special notes for your reviewer**:
- I added the `github.com/andybalholm/brotli` dependency for `br` reading/writing. It's the most used pure go library for this, and it's MIT licensed. However, I'm not sure where the license should be added in the Grafana repo.
- The go-cache library I added was already being used in repo.

### CODE REVIEW

**Review by andresmgot — COMMENTED** [review]
Great job here! One thing that I am unsure of is the use of the cache. That can lead to unexpected bugs. For example, if a user gets a forbidden error and adds the required permissions, they will still get the same error until the cache expires. Also, as far as I can see, the cache is only based on the URL of the request, what if now (or in the future) a resource requires something in the request body? cc/ @sunker in case I am missing some context here.

**Comment by iwysiu** [comment]
Added a comment to the regex!

I rebased to try to get rid of some local compilation errors, and the frontend test I removed was testing logic that's now in the backend.

**Comment by marefr** [comment]
Guess I have some questions about the cache as well. Why do you need a cache in the first place?

**Comment by iwysiu** [comment]
The cache is honestly because the original code was caching in the frontend (in api.ts), so I assumed there was significant enough traffic to warrant it. I'm ok with removing it.

**Comment by andresmgot** [comment]
> The cache is honestly because the original code was caching in the frontend (in api.ts), so I assumed there was significant enough traffic to warrant it. I'm ok with removing it.

Got it, one difference is that before (in api.ts), only successful requests were being cached, now we would be caching all. I would still avoid caching to avoid issues but there may be some cost-related issue I am not aware of. @sunker can you jump in here? You have more context. Thanks!

**Review by sunker — COMMENTED** [review]
Great work @iwysiu!! So happy to see the datasource.ts file cleaned up, and nice that you've moved the test datasource check to the backend too. 

>Guess I have some questions about the cache as well. Why do you need a cache in the first place?

Good question. Think it might be legacy from when this data source used to be in angular. 

Looking at the network traffic, it seems that projects are loaded three times when the query editor is rendered. I believe user pay per call to these api:s, so the cache does indeed have a purpose. However, the right way to solve this would be to prevent additional calls to get projects from within the react components. I think changing the Project.tsx file to something like this should do the trick:

```typescript
const [projects, setProjects] = useState<Array<SelectableValue<string>>>([]);
  useEffect(() => {
    datasource.getProjects().then((projects) => setProjects(projects));
  }, [datasource]);

  const projectsWithTemplateVariables = useMemo(
    () => [
      projects,
      {
        label: 'Template Variables',
        options: templateVariableOptions,
      },
      ...projects,
    ],
    [projects, templateVariableOptions]
  );
```

Once that is done, it should be fine to remove the cache completely.

File: pkg/tsdb/cloudmonitoring/resource_handler.go [hunk 1]
```diff
@@ -1,94 +1,296 @@
...
 	"net/url"
+	"regexp"
 	"strings"
 
+	"github.com/andybalholm/brotli"
+	"github.com/grafana/grafana-google-sdk-go/pkg/utils"
 	"github.com/grafana/grafana-plugin-sdk-go/backend/resource/httpadapter"
 )
 
+var nameExp = regexp.MustCompile(`([^\/]*)\/*$`)
```
[the remark this answers is not in the document]
Comment (iwysiu) [reply pkg/tsdb/cloudmonitoring/resource_handler.go:22]: done

File: pkg/tsdb/cloudmonitoring/resource_handler.go [hunk 2]
```diff
@@ -1,94 +1,297 @@
...
+
+type processResponse func(body []byte) ([]byte, error)
+
 func (s *Service) registerRoutes(mux *http.ServeMux) {
-	mux.HandleFunc("/cloudmonitoring/", s.resourceHandler(cloudMonitor))
-	mux.HandleFunc("/cloudresourcemanager/", s.resourceHandler(resourceManager))
+	mux.HandleFunc("/gceDefaultProject", getGCEDefaultProject)
+
+	mux.HandleFunc("/metricDescriptors/", s.resourceHandler(cloudMonitor, processMetricDescriptors))
+	mux.HandleFunc("/services/", s.resourceHandler(cloudMonitor, processServices))
```
[the remark this answers is not in the document]
Comment (iwysiu) [reply pkg/tsdb/cloudmonitoring/resource_handler.go:32]: done

File: pkg/tsdb/cloudmonitoring/resource_handler.go [hunk 3]
```diff
@@ -1,94 +1,297 @@
...
+	mux.HandleFunc("/slos/", s.resourceHandler(cloudMonitor, processSLOs))
+	mux.HandleFunc("/cloudresourcemanager/", s.resourceHandler(resourceManager, processProjects))
```
[the remark this answers is not in the document]
Comment (iwysiu) [reply pkg/tsdb/cloudmonitoring/resource_handler.go]: done

**Review by iwysiu — COMMENTED** [review]
Removed the cache and did a force push to resolve a merge conflict.

**Review by andresmgot — APPROVED** [review]
LGTM!

**Review by sunker — APPROVED** [review]
Great stuff @iwysiu! 

When testing this I did find a problem with annotations but that's not related to your changes so I'll create a separate issue for that.

**Pull request merged**